File: Exercism Codes/Java/Say.java
public class Say {
	private String[] onesDigit;
	private String[] tensDigit;
	private String[] teensDigit;
	
	public Say() {
		onesDigit = new String[] {
				"", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
		};
		tensDigit = new String[] {
				"", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
		};
		teensDigit = new String[] {
				"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
		};
	}
	
    public String say(long number) {
        String numberString = String.valueOf(number);
        String numberLetter = "";
        if(number == 0) {
        	numberLetter = "zero";
        }
        else if(number < 0 || number > 999999999999L) {
        	throw new IllegalArgumentException();
        }
        else {
        	for(int i = numberString.length() - 1; i >= 0; i--) {
            	int checkIndex = (numberString.length() - 1) - i;
        		int index = Character.getNumericValue(numberString.charAt(i));
            	switch(checkIndex % 3) {
            	case 0:
            		String thousands = " ";
                	switch(checkIndex) {
                	case 3:
                		thousands = CheckThousands(numberString, checkIndex, 7, " thousand ");
                		break;
                	case 6:
                		thousands = CheckThousands(numberString, checkIndex, 10, " million ");
                		break;
                	case 9:
                		thousands = CheckThousands(numberString, checkIndex, 13, " billion ");
                		break;
                	case 12:
                		thousands = CheckThousands(numberString, checkIndex, 16, " trillion ");
                		break;
                	}
                	if(i - 1 >= 0) {
                    	if(numberString.charAt(i - 1) != '1') {
                        	numberLetter = onesDigit[index] + thousands + numberLetter;
                    	}
                    	else {
                			numberLetter = teensDigit[index] + thousands + numberLetter;
                		}
                	}
                	else {
                		numberLetter = onesDigit[index] + thousands + numberLetter;
                	}
                	break;
            	case 1:
            		if(numberString.charAt(i) != '1' && numberString.charAt(i) != '0') {
            			if(numberString.charAt(i + 1) != '0') {
            				numberLetter = tensDigit[index] + "-" + numberLetter;
            			}
            			else {
            				numberLetter = tensDigit[index] + numberLetter;
            			}
            		}
                	break;
            	case 2:
            		if(numberString.charAt(i) != '0') {
                		if(numberString.charAt(i + 1) != '0' || numberString.charAt(i + 2) != '0') {
                			numberLetter = onesDigit[index] + " hundred " + numberLetter;
            			}
            			else {
            				numberLetter = onesDigit[index] + " hundred" + numberLetter;
            			}
            		}
                	break;
            	}
            }
        }
        
        return numberLetter.trim();
    }
    
    public String CheckThousands(String numberString, int index, int length, String name) {
    	if(numberString.length() >= length) {
    		if(!(numberString.charAt(index) == '0' && numberString.charAt(index + 1) == '0' && numberString.charAt(index + 2) == '0')) {
    			return name;
    		}
		}
		else {
			return name;
		}
    	return "";
    }
}

File: Exercism Codes/JavaScript/say.js
const ones = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
const tens = ["ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];
const teens = ["eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"];

export const say = (n) => {
  if (n === 0) {
    return "zero";
  }
  if (n < 0 || n > 999999999999) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  const result = [];
  const str = String(n);

  for (let x = str.length - 1; x >= 0; x--) {
    const idx = str.length - x - 1;
    if (idx / 3 === 1) {
      result.unshift("thousand");
    }
    if (idx / 3 === 2) {
      if (parseInt(str.substring(x + 1, x + 4)) === 0) {
        result.shift();
      }
      result.unshift("million");
    }
    if (idx / 3 === 3) {
      if (parseInt(str.substring(x + 1, x + 4)) === 0) {
        result.shift();
      }
      result.unshift("billion");
    }

    if (parseInt(str[x]) !== 0) {
      if (idx % 3 === 0) {
        result.unshift(ones[parseInt(str[x]) - 1]);
      }
      if (idx % 3 === 1) {
        if (parseInt(str[x]) === 1 && parseInt(str[x + 1]) !== 0) {
          result.shift();
          result.unshift(teens[parseInt(str[x + 1]) - 1]);
        }
        else {
          result.unshift(tens[parseInt(str[x]) - 1]);
        }
      }
      if (idx % 3 === 2) {
        result.unshift(ones[parseInt(str[x]) - 1], "hundred");
      }
    }
  }

  return join(result);
};

const join = (lst) => {
  let result = lst[0];

  for (let x = 1; x < lst.length; x++) {
    if (tens.includes(lst[x - 1]) && ones.includes(lst[x])) {
      result += "-" + lst[x];
    }
    else {
      result += " " + lst[x];
    }
  }

  return result;
};